### ISSUE TITLE
set_pin is not invoked when registering a security key with UV = required and pin_auth = None

### ISSUE THREAD

**Issue by andreydanil** [mozilla/authenticator-rs#223]
In Chrome, in the case where a security token does not have a FIDO2 pin set, with user verification set to `required`, the browser prompts the user to set the security token PIN during registration.

## Chrome experience:
1. Reset FIDO2 on the security key.
2. Go to https://webauthn.io/.
3. Expand the advanced settings, and select User Verification as `Required`. 
4. Enter `jondoe` as a username in the `example_username` field.
5. Click on `Register`. 
6. Examine the prompt to set the FIDO2 pin.
7. Tap on the token to complete the registration.
<img width="1174" alt="chrome-webauthn" src="https://user-images.githubusercontent.com/5439988/215604423-50da6c30-d432-4a85-82e9-5c86e0893a16.png">

## Firefox bug:
1. Reset FIDO2 on the security key.
2. Go to https://webauthn.io/.
3. Expand the advanced settings, and select User Verification as `Required`. 
4. Enter `jondoe` as  a username in the `example_username` field.
5. Click on `Register`. 
6. There is an error: `The authenticator was unable to process the specified options, or could not create a new credential`.
7. The registration fails.
![ff-nightly](https://user-images.githubusercontent.com/5439988/215606763-a5773dc9-b656-43ea-bb4a-4714e56a3b5a.png)

Chrome makes this type of WebAuthn request:
```
{1: ["FIDO_2_0", "FIDO_2_1_PRE"], 2: ["credProtect", "hmac-secret"], 3: h'2FC0579F811347EAB116BB5A8DB9202A', 4: {"rk": true, "up": true, "plat": false, "clientPin": false, "credentialMgmtPreview": true}, 5: 1200, 6: [2, 1], 7: 8, 8: 128, 9: ["usb"], 10: [{"alg": -7, "type": "public-key"}, {"alg": -8, "type": "public-key"}], 13: 4, 14: 328707}
```
Which then causes the Chrome Authenticator library to set the pin.

Is there a reason why `UV = required` is not prompting users to set the FIDO2 PIN in the Firefox implementation?

**Comment by msirringhaus** [comment]
Thanks for reporting! I'm aware of this. The reason for not having it in Firefox as of now is simply, that the UI for it hasn't been implemented yet. This crate is capable of setting (and changing) the PIN. But the functionality hasn't yet been exposed to Firefox.

**Comment by andreydanil** [comment]
Thanks for the reply [msirringhaus](https://github.com/msirringhaus).
Just curious which repository contains the UI? I know FF uses Mercurial so I was only able to find forks on GitHub. Perhaps we can do a deep dive and implement it.
 
I did a GitHub search and the set pin appears to be missing from the FF source (see this fork): https://github.com/bolucat/Firefox/blob/eadb2ecc5bca616861f03bf7e575592b210af37e/browser/locales/en-US/browser/webauthnDialog.ftl#L13. There are WebAuthn prompts for `webauthn-pin-required-prompt`  and `webauthn-pin-invalid-prompt`, but not to set the pin prompt. 

I see that you have implemented the set pin here: https://github.com/mozilla/authenticator-rs/blob/ctap2-2021/examples/set_pin.rs, but was curious why it was missing during registration. We can keep this bug open for tracking this issue. Thoughts?

**Comment by msirringhaus** [comment]
The sources of FF can be found [here](https://hg.mozilla.org/mozilla-central/), although a more searchable online-version can be found [here](https://searchfox.org/mozilla-central/source/).

One of the reasons, "PIN required" is implemented, but "Set PIN" is not, is simply that the usage of the former is basically required to have FIDO work, the latter is more "nice to have". Firefox uses the tokens, but isn't (as a first step) expected to manage them as well. Although it certainly would be nice, if it could.

The other reason is that the necessary code for the UI-dialog for a simple PIN-entry was already there, and could be reused. A proper "Set PIN"-dialog would need to look somewhat different (as your screenshot of Chromium shows). This is a) more work and b) requires the involvement of the UX- and design-teams. 

I do hope that we can get this feature in the future, though.

**Comment by msirringhaus** [comment]
Ah, thanks for pointing out the management page!

I'd say, whether we should have this dialog or not is best discussed in the bugzilla-ticket, since it has nothing to do with this crate. This crate is capable of doing it. This is purely a Firefox-UX decision (and implementation), hence we will also need to involve the UX-team.

**Comment by andreydanil** [comment]
Following up on this [msirringhaus](https://github.com/msirringhaus).

FF has switched to the macOS popup for the security key interactions - which utilizes the platform APIs. Is this bug still blocked on setting the PIN in side the webauthn registration `navigator.credentials.create` API, since the FF UI is no longer the blocker?

Authentication Webauthn prompt, in Firefox 125:
<img width="961" alt="image" src="https://github.com/mozilla/authenticator-rs/assets/5439988/40b85acc-1897-4160-9a2b-addb08a40a46">

**Comment by micolous** [comment]
> FF has switched to the macOS popup for the security key interactions - which utilizes the platform APIs. Is this bug still blocked on setting the PIN in side the webauthn registration `navigator.credentials.create` API, since the FF UI is no longer the blocker?

Firefox uses platform-native WebAuthn APIs on Android, macOS and Windows to communicate with authenticators, so those operations don't touch this library. The platform APIs prompt the user to set a PIN when it is needed, with their own UI. They only sends Firefox the response from the authenticator, so it's all the same regardless of when UV was configured (aside from it being slower).

On iOS, Firefox uses Safari's rendering engine (like all third-party browsers).

For other platforms, it's still blocked on adding Firefox UI.
